modules/objdetect/src/qrcode.cpp

```diff
@@ -967,1407 +3808,2407 @@ bool detectQRCode(InputArray in, vector<Point> &points, double eps_x, double eps
+        /**calcutlate the total data codewords of current version level*/
+        data_codewords = version_info_database[i].total_codewords
+                             -
+                             tmp_ecc_params->ecc_codewords*(tmp_ecc_params->num_blocks_in_G1+tmp_ecc_params->num_blocks_in_G2);
+        /**find the minimum version which can hold all the bit stream */
+        if(data_codewords*byte_len > input_length ){
+            version_index = i;
+            break;
+        }
+    }
+    return version_index;
+}
+
+/**
+ *
+ * */
+bool QREncoder::versionEstimate(const int &input_length,vector<int>&possible_version){
+    possible_version.clear();
+    if(input_length > version_capacity_database[40].ec_level[ecc_level].encoding_modes[1])
+        return false;
+    if(input_length <= version_capacity_database[9].ec_level[ecc_level].encoding_modes[3]){
+        possible_version.push_back(1);
+    }
+    else if (input_length <= version_capacity_database[9].ec_level[ecc_level].encoding_modes[1]){
+        possible_version.push_back(1);
+        possible_version.push_back(2);
+    }
+    else if(input_length <= version_capacity_database[26].ec_level[ecc_level].encoding_modes[3]){
+        possible_version.push_back(2);
+    }
+    else if(input_length <= version_capacity_database[26].ec_level[ecc_level].encoding_modes[1]){
+        possible_version.push_back(2);
+        possible_version.push_back(3);
+    }
+    else{
+        possible_version.push_back(3);
+    }
+    return true;
+}
+
+int QREncoder::versionAuto(const std::string & input_str){
+    vector<int> possible_version;
+    /**first try to estimate the version range (0-9 , 10-26, 27-41)*/
+    versionEstimate((int)input_str.length(),possible_version);
+    int v = 0;
+    vector<uint8_t> payload_tmp;
+    /**the beginning index of version */
+    int version_range[5]={0,1,10,27,41};
+    for(size_t i = 0 ; i < possible_version.size() ; i++ ){
+        int version_range_index = possible_version[i];
+        if(version_range_index == 1 ){
+            v = 1;
+        }
+        else if (version_range_index == 2 ){
+            v = 10;
+        }
+        else{
+            v = 27;
+        }
+        /**use the output bit stream to find the exact version level*/
+        encodeAuto(input_str,payload_tmp);
+        v = findVersionCapacity((int)payload_tmp.size(),ecc_level,
+                                            version_range[version_range_index],version_range[version_range_index+1]);
+        /**find then break*/
+        if(v!= -1)
+            break;
     }
-    if (true_points_group.size() == 0)
-        return false;
-    if (true_points_group[0].size() < 3)
-        return false;
+    return v;
+}
 
+QREncoder::QREncoder(const std::string& input ,int mode,int v  = 0 ,int ecc = 0 ,int mask = -1 ,int eci_mode = -1,int structure_num = 2){
```

Comment:
create `QREncoder::init` method (as it implemented for classes `QRDetect` and `QRDecode` )